openstacksdk_plugin/resources/network/port.py

```diff
@@ -15,33 +15,170 @@
+    create the port instance using Openstack API
+    """
+
+    # Get security groups from port config
+    security_groups = port_config.get('security_groups')
+
+    # Get the security groups from relationship if any
+    rel_security_groups = \
+        find_openstack_ids_of_connected_nodes_by_openstack_type(
+            ctx, SECURITY_GROUP_OPENSTACK_TYPE)
+
+    # Check if network config comes from two sources or not
+    if rel_security_groups and security_groups:
+        raise NonRecoverableError('Port can\'t both have the '
+                                  '"security_groups" property and be '
+                                  'connected to a network via a '
+                                  'relationship at the same time')
+
+    port_config['security_groups'] = security_groups or rel_security_groups
 
 
 def _update_port_config(port_config):
     """
     This method will try to resolve if there are any nodes connected to the
     port node and try to update the configurations from nodes in order to
     help create port from configurations
-    :param port_config:
+    :param port_config: The port configuration required in order to
+    create the port instance using Openstack API
     """
 
+    # Update network config for port node
+    _update_network_config(port_config)
 
-@with_openstack_resource(OpenstackPort)
+    # Update security groups config for port node
+    _update_security_groups_config(port_config)
+
+
+def _update_external_port(openstack_resource):
+    """
+    This method will update external port by attaching new ips to external
+    port
+    :param openstack_resource: Instance Of OpenstackPort in order to
+    use it
+    """
+    # Get the external port using the resource id provided via port node
+    external_port = openstack_resource.get()
+    # Check if the current port node has allowed_address_pairs as part of
+    # resource_config
+    addresses_to_add = openstack_resource.config.get('allowed_address_pairs')
+    if addresses_to_add:
+        old_addresses = external_port.get('allowed_address_pairs') or []
+
+        # Get the old ips from the each pair
+        old_ips = \
+            [
+                old_address['ip_address']
+                for old_address
+                in old_addresses if old_address.get('ip_address')
+            ]
+        # Get the ips need to be added to the external port
+        ips_to_add = \
+            [
+                address_to_add['ip_address']
+                for address_to_add
+                in addresses_to_add if address_to_add.get('ip_address')
+            ]
+
+        # Check if there are a common ips between old ips and the one we
+        # should add via node
+        common_ips = set(old_ips) & set(ips_to_add)
+        if common_ips:
+            raise NonRecoverableError(
+                'Ips {0} are already assigned to {1}'
+                ''.format(common_ips, external_port.id))
```

Comment:
Is this from the old plugin? Should we add the ability to skip it if it is already in the config?

Reply:
This is how the old plugin handle it, if the ips are already assigned to the port and you are trying to assign them again it raises error